download/oci_dowload.go

```diff
@@ -0,0 +1,366 @@
+	}
+	//currently it has 3 as it has the manifest, tar and config layers
+	if len(layers) != 3 {
+		return nil, fmt.Errorf("expected 3 layers only")
+	}
+
+	tarballDescriptor := ocispec.Descriptor{}
+	for i := range layers {
+		if layers[i].MediaType == "application/vnd.oci.image.layer.v1.tar+gzip" {
+			tarballDescriptor = layers[i]
+			break
+		}
+	}
+	if tarballDescriptor.MediaType == "" {
+		return nil, fmt.Errorf("no tarball descriptor found in the layers")
+	}
+
+	bundleFilePath := filepath.Join(d.localStorePath, "blobs", "sha256", string(tarballDescriptor.Digest.Hex()))
+	fileReader, err := os.Open(bundleFilePath)
+	if err != nil {
+		return nil, err
+	}
+	loader := bundle.NewTarballLoaderWithBaseURL(fileReader, d.localStorePath)
+	reader := bundle.NewCustomReader(loader).WithBaseDir(d.localStorePath)
+	bundleInfo, err := reader.Read()
+	if err != nil {
+		return &downloaderResponse{}, fmt.Errorf("unexpected error %w", err)
+	}
+
+	m.Timer(metrics.BundleRequest).Stop()
+
+	return &downloaderResponse{
+		b:        &bundleInfo,
+		raw:      fileReader,
+		etag:     "",
+		longPoll: false,
+	}, nil
+}
+
+func (d *OCIDownload) pull(ref string) (*ocispec.Descriptor, []ocispec.Descriptor, error) {
+	clientConfig := d.client.Config()
+	authHeader := make(http.Header)
+	resolver := docker.NewResolver(docker.ResolverOptions{
+		Hosts: func(s string) ([]docker.RegistryHost, error) {
+			var client *http.Client
+
+			var err error
+			if clientConfig != nil && clientConfig.Credentials.ClientTLS != nil {
+				client, err = clientConfig.Credentials.ClientTLS.NewClient(*clientConfig)
+				if err != nil {
+					return nil, fmt.Errorf("can not create a new client: %w", err)
+				}
+			} else {
+				if clientConfig != nil && clientConfig.Credentials.Bearer != nil {
+					client, err = d.client.Config().Credentials.Bearer.NewClient(*clientConfig)
+					if err != nil {
+						return nil, fmt.Errorf("can not create a new bearer client: %w", err)
+					}
+					//Attach auth header
+					authHeader.Add("Authorization",
+						fmt.Sprintf("%s %s",
+							clientConfig.Credentials.Bearer.Scheme,
+							base64.StdEncoding.EncodeToString([]byte(clientConfig.Credentials.Bearer.Token))),
+					)
+				} else {
+					client = &http.Client{Transport: &http.Transport{TLSClientConfig: &tls.Config{InsecureSkipVerify: clientConfig.AllowInsecureTLS}}}
+				}
+			}
+			urlInfo, err := url.Parse(clientConfig.URL)
+			if err != nil {
+				return nil, fmt.Errorf("invalid host url: %s error:%w", clientConfig.URL, err)
+			}
+			return []docker.RegistryHost{
+				{
+					Host:         urlInfo.Host,
+					Scheme:       urlInfo.Scheme,
+					Capabilities: docker.HostCapabilityPull | docker.HostCapabilityResolve | docker.HostCapabilityPush,
+					Client:       client,
+					Path:         "/v2",
+					Authorizer: docker.NewDockerAuthorizer(
+						docker.WithAuthClient(client),
+						docker.WithAuthCreds(func(s string) (string, string, error) {
```

Comment:
[nit] looks like it's unused?

```suggestion
						docker.WithAuthCreds(func(string) (string, string, error) {
```